src/ComputeSharp.D2D1.CodeFixers/MissingPixelShaderDescriptorOnPixelShaderCodeFixer.cs

```diff
@@ -0,0 +1,103 @@
+        {
+            // Register the code fix to update the return type to be Task instead
+            context.RegisterCodeFix(
+                CodeAction.Create(
+                    title: "Add [D2DGeneratedPixelShaderDescriptor] attribute",
+                    createChangedDocument: token => ChangeReturnType(context.Document, root, structDeclaration),
+                    equivalenceKey: "Add [D2DGeneratedPixelShaderDescriptor] attribute"),
+                diagnostic);
+        }
+    }
+
+    /// <summary>
+    /// Applies the code fix to add the [D2DGeneratedPixelShaderDescriptor] attribute to a target type.
+    /// </summary>
+    /// <param name="document">The original document being fixed.</param>
+    /// <param name="root">The original tree root belonging to the current document.</param>
+    /// <param name="structDeclaration">The <see cref="StructDeclarationSyntax"/> to update.</param>
+    /// <returns>An updated document with the applied code fix, and the return type of the method being <see cref="Task"/>.</returns>
+    private static Task<Document> ChangeReturnType(Document document, SyntaxNode root, StructDeclarationSyntax structDeclaration)
+    {
+        int index = 0;
+
+        // Find the index to use to insert the attribute. We want to make it so that if the struct declaration
+        // has a bunch of D2D attributes, the new one will be inserted right after that. This way the final list
+        // will be nicely sorted, instead of having D2D attributes interleaving other unrelated attributes, if any.
+        foreach (AttributeListSyntax attributeList in structDeclaration.AttributeLists)
+        {
+            // Make sure we have an attribute to check
+            if (attributeList.Attributes is not [AttributeSyntax attribute, ..])
+            {
+                continue;
+            }
+
+            // Make sure we find a readable identifier for the attribute
+            if (attribute.Name is not IdentifierNameSyntax { Identifier.Value: string identifier })
+            {
+                break;
+            }
+
+            // If the attribute is D2D one, increment the index and continue
+            if (identifier.Contains("D2D"))
+            {
+                index++;
+            }
+            else
+            {
+                // Otherwise, stop here, we reached the end of the sequence
+                break;
+            }
+        }
+
+        // Create the attribute syntax for the new attribute
+        AttributeListSyntax newAttributeList = AttributeList(SingletonSeparatedList(Attribute(IdentifierName("D2DGeneratedPixelShaderDescriptor"))));
+
+        // Create a new syntax node with the new attribute
+        SyntaxNode typeSyntax = SyntaxGenerator.GetGenerator(document).InsertAttributes(structDeclaration, index, newAttributeList);
+
+        // Replace the node in the document tree
+        Document updatedDocument = document.WithSyntaxRoot(root.ReplaceNode(structDeclaration, typeSyntax));
```

Comment:
I think this is assuming that the namespace containing the attribute is already imported.

Reply:
I thought I could rely on that given that in order for this error to appear, the type must be implementing `ID2D1PixelShader`, which is in the same namespace 🤔

I suppose I could use that special annotation to add using directives if needed, and that'd only add that if actually needed right?

Reply:
Well, it's not guaranteed for two cases:

- A user could be having `partial struct S : ISomeInterface` in one file, and `partial struct S` in another file. In this case, the diagnostic location is not known. It could end up in either partials, depending on the order of `.cs` files passed to the compiler.
- A user could be having `struct S : global::Fully.Qualified.Name.Of.ISomeInterface`

These are edge cases that you can choose not to handle, which could be acceptable.

The annotation will add it only if needed, yes. Otherwise, it would be a Roslyn bug.